cfme/networks/provider/__init__.py

```diff
@@ -144,6 +146,76 @@ def create(self, cancel=False, validate_credentials=True, validate_inventory=Fal
 
         return created
 
+    def update(self, updates, cancel=False, validate_credentials=True):
+        """
+        Updates a provider in the UI.  Better to use utils.update.update context
+        manager than call this directly.
+
+        Args:
+           updates (dict): fields that are changing.
+           cancel (boolean): whether to cancel out of the update.
+           validate_credentials (boolean): whether credentials have to be validated
+        """
+        edit_view = navigate_to(self, 'Edit')
+
+        # filling main part of dialog
+        endpoints = updates.pop('endpoints', None)
+        if updates:
+            edit_view.fill(updates)
+
+        # filling endpoints
+        if endpoints:
+            endpoints = self._prepare_endpoints(endpoints)
+
+            for endpoint in endpoints.values():
+                # every endpoint class has name like 'default', 'events', etc.
+                # endpoints view can have multiple tabs, the code below tries
+                # to find right tab by passing endpoint name to endpoints view
+                try:
+                    endp_view = getattr(self.endpoints_form(parent=edit_view), endpoint.name)
+                except AttributeError:
+                    # tabs are absent in UI when there is only single (default) endpoint
+                    endp_view = self.endpoints_form(parent=edit_view)
+                endp_view.fill(endpoint.view_value_mapping)
+
+                # filling credentials
+                # the code below looks for existing endpoint equal to passed one and
+                # compares their credentials. it fills passed credentials
+                # if credentials are different
+                cur_endpoint = self.endpoints[endpoint.name]
+                if hasattr(endpoint, 'credentials'):
+                    if not hasattr(cur_endpoint, 'credentials') or \
+                            endpoint.credentials != cur_endpoint.credentials:
+                        if hasattr(endp_view, 'change_password'):
+                            endp_view.change_password.click()
+                        elif hasattr(endp_view, 'change_key'):
+                            endp_view.change_key.click()
+                        else:
+                            NotImplementedError(
+                                "Such endpoint doesn't have change password/key button")
+
+                        endp_view.fill(endpoint.credentials.view_value_mapping)
+                if (validate_credentials and hasattr(endp_view, 'validate') and
+                        endp_view.validate.is_displayed):
+                    endp_view.validate.click()
+                    self._post_validate_checks(edit_view)
+
+        details_view = self.create_view(navigator.get_class(self, 'Details').VIEW)
+        main_view = self.create_view(navigator.get_class(self, 'All').VIEW)
+
+        if cancel:
+            edit_view.cancel.click()
```

Comment:
Why not just have a _post_cance() routine and include the cancel click with the checks.   It seems kind of odd the separation.